File: packages/swt3-ai-dotnet/Signing.cs
using System.Security.Cryptography;
using System.Text;

namespace Swt3Ai;

/// <summary>
/// HMAC-SHA256 payload signing for non-repudiation.
/// </summary>
public static class Signing
{
    /// <summary>
    /// Sign a payload with HMAC-SHA256.
    /// If agentId is provided, the message is "{fingerprint}:{agentId}".
    /// Otherwise, the message is just the fingerprint.
    /// Returns a 64-character lowercase hex string.
    /// </summary>
    public static string SignPayload(
        string signingKey,
        string anchorFingerprint,
        string? agentId = null)
    {
        var message = agentId != null
            ? $"{anchorFingerprint}:{agentId}"
            : anchorFingerprint;

        var keyBytes = Encoding.UTF8.GetBytes(signingKey);
        var messageBytes = Encoding.UTF8.GetBytes(message);

        var hash = HMACSHA256.HashData(keyBytes, messageBytes);
        return Convert.ToHexString(hash).ToLowerInvariant();
    }
}

File: packages/swt3-ai-ruby/lib/swt3_ai/signing.rb
require "openssl"

module Swt3Ai
  module Signing
    # Sign a payload with HMAC-SHA256 for non-repudiation.
    #
    # If agent_id is provided, the message is "{fingerprint}:{agent_id}".
    # Otherwise, the message is just the fingerprint.
    # Returns a 64-character lowercase hex string.
    def self.sign_payload(signing_key, anchor_fingerprint, agent_id = nil)
      message = agent_id ? "#{anchor_fingerprint}:#{agent_id}" : anchor_fingerprint
      OpenSSL::HMAC.hexdigest("SHA256", signing_key, message)
    end
  end
end
